File: leetcode/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int n = s.size();
        int ans = 0;
        int j = 0;
        int i = 0;
        unordered_set<char> set;

        while (i < n && j < n) {
            if (set.find(s[j]) == set.end()) {
                set.insert(s[j++]);
            } else {
                set.erase(s[i++]);
            }
            ans = max(ans, j-i);
        }
        return ans;
    }
};

File: leetcode/3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        left, output = 0, 0
        seen = {}
        for right, curr in enumerate(s):
            if curr in seen:
                left = max(left, seen[curr] + 1)
            output = max(output, right - left + 1)
            seen[curr] = right
        return output
        